Sofa/Component/Constraint/Lagrangian/Model/src/sofa/component/constraint/lagrangian/model/BaseContactLagrangianConstraint.h

```diff
@@ -0,0 +1,160 @@
+#include <sofa/core/behavior/MechanicalState.h>
+#include <iostream>
+#include <map>
+#include <deque>
+#include <type_traits>
+#include <concepts>
+
+namespace sofa::component::constraint::lagrangian::model
+{
+
+
+//TODO(Paul) C++20: add concept for ContactParams
+// Every ContactParams should implement hasTangentialComponent
+struct BaseContactParams
+{
+    virtual bool hasTangentialComponent() const = 0;
+};
+
+template<class DataTypes, class ContactParams>
+class BaseContactLagrangianConstraint : public core::behavior::PairInteractionConstraint<DataTypes>
+{
+public:
+    SOFA_CLASS(SOFA_TEMPLATE2(BaseContactLagrangianConstraint,DataTypes,ContactParams), SOFA_TEMPLATE(core::behavior::PairInteractionConstraint,DataTypes));
+
+    typedef typename DataTypes::VecCoord VecCoord;
+    typedef typename DataTypes::VecDeriv VecDeriv;
+    typedef typename DataTypes::MatrixDeriv MatrixDeriv;
+    typedef typename DataTypes::MatrixDeriv::RowConstIterator MatrixDerivRowConstIterator;
+    typedef typename DataTypes::MatrixDeriv::ColConstIterator MatrixDerivColConstIterator;
+    typedef typename DataTypes::MatrixDeriv::RowIterator MatrixDerivRowIterator;
+    typedef typename DataTypes::MatrixDeriv::ColIterator MatrixDerivColIterator;
+    typedef typename DataTypes::Coord Coord;
+    typedef typename DataTypes::Deriv Deriv;
+    typedef typename Coord::value_type Real;
+    typedef typename core::behavior::MechanicalState<DataTypes> MechanicalState;
+
+    typedef core::behavior::BaseConstraint::ConstraintBlockInfo ConstraintBlockInfo;
+    typedef core::behavior::BaseConstraint::PersistentID PersistentID;
+    typedef core::behavior::BaseConstraint::ConstCoord ConstCoord;
+
+    typedef core::behavior::BaseConstraint::VecConstraintBlockInfo VecConstraintBlockInfo;
+    typedef core::behavior::BaseConstraint::VecPersistentID VecPersistentID;
+    typedef core::behavior::BaseConstraint::VecConstCoord VecConstCoord;
+    typedef core::behavior::BaseConstraint::VecConstDeriv VecConstDeriv;
+    typedef core::behavior::BaseConstraint::VecConstArea VecConstArea;
+
+    typedef core::objectmodel::Data<VecCoord>		DataVecCoord;
+    typedef core::objectmodel::Data<VecDeriv>		DataVecDeriv;
+    typedef core::objectmodel::Data<MatrixDeriv>    DataMatrixDeriv;
+
+    typedef typename core::behavior::PairInteractionConstraint<DataTypes> Inherit;
+
+protected:
+
+    struct Contact
+    {
+        int m1, m2;		///< the two extremities of the spring: masses m1 and m2
+        Deriv norm;		///< contact normal, from m1 to m2
+        Deriv t;		///< added for friction
+        Deriv s;		///< added for friction
+        Real contactDistance;
+
+        unsigned int id;
+        long contactId;
+        PersistentID localId;
+        ContactParams parameters;
+
+        Coord P, Q;
+
+        mutable Real dfree;
+    };
+
+    sofa::type::vector<Contact> contacts;
+    bool yetIntegrated;
+    SReal customTolerance;
+
+    bool* contactsStatus;
```

Comment:
I am sure we can do better than a raw pointer...

Reply:
Actually, it seems like this is used nowhere. I suggest to keep it in this PR and make another one after one removing this vector and its usage in the codebase.